Add vqa

henryj18 opened the pull request:
This PR includes changes on image-caption and vqa code based off add_text_modality

sidhsikka1 commented:
Moved Discord CTA up in Contributing guide, made prose project specific, and removed excess verbage.

henryj18 committed:
Merge remote-tracking branch 'origin/add_text_modality' into add_image_caption

sidhsikka1 committed:
Update README.md

Added new google drive links

harshsikka committed:
Update README.md

harshsikka committed:
Merge pull request #26 from ManifoldRG/harshsikka-patch-1

Update README.md

henryj18 committed:
Commit in order to pull in the up to date changes from parent branch

henryj18 committed:
Merge remote-tracking branch 'origin/add_text_modality' into add_image_caption

henryj18 committed:
minor

henryj18 committed:
Change per test

henryj18 committed:
Change per test

henryj18 committed:
Add "to(self.device)"

henryj18 committed:
fix some issues per test

henryj18 committed:
Fix the problem of logits cut short

henryj18 committed:
This is the Notebook for testing image-caption task

henryj18 committed:
Minor changes: to remove the cell outputs

henryj18 committed:
Add vqa task

henryj18 committed:
minor change

henryj18 committed:
VQA test script

henryj18 committed:
Change on data pipelines

henryj18 committed:
re-structure the data pipeline for vqa task

henryj18 committed:
remove default values from VqaTask's  __init___

henryj18 committed:
Update to include image-caption task and VQA task

henryj18 committed:
fix some formatting

henryj18 committed:
minor wording changes

henryj18 committed:
Ass VQA dataset and polishing some wording

henryj18 committed:
minor change of indentation

henryj18 committed:
More wording and indentation change

henryj18 committed:
Mention the test script on Colab

henryj18 committed:
Add more description about VQA data

henryj18 committed:
Add comments about contributing new code

henryj18 committed:
Minor wording change

henryj18 committed:
Add reference to an example PR

henryj18 committed:
Correct the reference of PR

henryj18 committed:
Minor fix of the PR reference

henryj18 committed:
Another minor fix of the PR reference

eihli committed:
fix broken link to DeepMind's GATO publication

deepmind.google removed their /publications endpoint.

They still have a blog post about GATO at https://deepmind.google/discover/blog/a-generalist-agent/ but the blog post doesn't link to the paper so I'm linking the InternetArchive's most recent version.

bhavul committed:
Merge pull request #1 from ManifoldRG/add_text_modality

Add text modality, and some basic refactoring

bhavul committed:
Merge pull request #32 from eihli/patch-1

fix broken link to DeepMind's GATO publication

henryj18 committed:
Merge branch 'master' into add_vqa

henryj18 committed:
Merge branch 'master' into add_vqa

henryj18 committed:
Revise Multiple_Tasks.ipynb per the most up to date changes

henryj18 committed:
Fix the proportion and number of samples of each task in a batch

henryj18 committed:
Uncomment the lines of code to clone git in Multiple_Tasks.ipynb

eihli committed:
Update TaskType references to use enum, not str

eihli committed:
Fix version dependency issue #54

Some of the HuggingFace libraries require typing_extensions >= 4.5.0 but Minari listed
it as an exact 4.4.0 requirement. Minari version 0.4.2 updates that exact requirement
to be a >= requirement.

bhavul commented on CONTRIBUTING.md:
```diff
@@ -16,4 +16,18 @@ See [Issues](https://github.com/ManifoldRG/NEKO/issues) for more information. If
+- Define the function for inference/predicton specific to this task in `gato/policy/gato_policy.py`. The rest of this file defines the training model for NEKO, and most likely you do not need to touch that part
+- Add a few lines of code in `gato/training/trainer.py` to sample data for the new task and add to each batch for training
+- Add a few lines of code in `train.py` to input the command-line args specific to your task and to instantiate an instance of your task
+
```
We will need another entry -- we should probably add an independent evaluation code in `eval.py`

henryj18 replied:
Sure, let's discuss about it, just came back from vacations, sorry for the delay

bhavul commented on gato/policy/gato_policy.py:
```diff
@@ -340,7 +347,9 @@ def tokenize_input_dicts(self, inputs: list):
                     tokens for tokens in
-                    [text_tokens, image_tokens, continuous_tokens, discrete_tokens, separator_tokens, continuous_action_tokens, discrete_action_tokens]
+                    # Swap the order of image_tokens and text_tokens (GATO paper places test tokens before image tokens) for the image-caption task,
+                    # caption (which a text_tokens) should follow image instead of preceding image, so that image data can auto-regressively predict caption tokens
```
I'm unsure if this is actually how it works. 
@daniellawson9999 care to comment?

bhavul commented on gato/policy/gato_policy.py:
```diff
@@ -458,6 +467,88 @@ def predict_text(self, batch_dict, max_length=20, deterministic=True):
+            # Note: the following 'text' is actually text token IDs instead of actual text. 
+            # We were using actual text before (and at that time tokenize_input_dicts(...) function also took actual text), and found an error 
+            # due to an edge case. The edge case is that the generated tokens could contain a sequence of .... (periods) or ,,,, (commas), 
+            # and if actual text is used in the following 'text', the tokenizer will treat the sequence as one token (which is not the desired behavior) 
```
The comments related to `text` can be removed. It wouldn't be helpful to someone in future.

bhavul commented on gato/policy/gato_policy.py:
```diff
@@ -458,6 +467,88 @@ def predict_text(self, batch_dict, max_length=20, deterministic=True):
+            #   n_patches - 1 + len(prompt_tokens) + idx is the index of the token in the sequence that will predict the next text token of the generated response
+            #   squeeze the logits from shape (1, 1, vocab_size) to shape (vocab_size) and restrict the choice to the range 
+            #       within text tokens. The complete vocab contains non-text toksns also, we need to to exclude them from the choice
+            next_token_logits = logits[:, n_patches -1 + len(prompt_tokens) + idx, :].squeeze()[:self.text_tokens]     # shape (self.text_tokens)
```
This is probably incorrect slicing. I will debug to check but it looks quite different from what's happening within text.

bhavul commented on gato/policy/gato_policy.py:
```diff
@@ -458,6 +467,88 @@ def predict_text(self, batch_dict, max_length=20, deterministic=True):
+            #       within text tokens. The complete vocab contains non-text toksns also, we need to to exclude them from the choice
+            next_token_logits = logits[:, n_patches -1 + len(prompt_tokens) + idx, :].squeeze()[:self.text_tokens]     # shape (self.text_tokens)
+            if deterministic:
+                next_token = torch.argmax(next_token_logits).item()
```
This may require `dim` as well

bhavul commented on gato/tasks/caption_task.py:
```diff
@@ -0,0 +1,174 @@
+        caption_dataset is the directory for all of the data (training and test)
+        train_data and test_data are list of sub_diretories under caption_dataset, with each directory containing 
+        one training or test dataset which is composed of multiple .tar files downloaded with img2dataset. 
+        Each tar file contains multiple bundles, with each bundle containing one .jpg, one txt and one json file. 
```
We should to have a better standard here, so as we scale up machines, it's easy to try a different dataset.

bhavul commented on gato/tasks/caption_task.py:
```diff
@@ -0,0 +1,174 @@
+if __name__ == '__main__':
+    # replace the following directory with your data directory
+    task = CaptionTask(task_type = 'caption', tokenizer_model = 'gpt2', caption_dataset = '/home/<user name>/Git/NEKO/Caption_data', 
+                       train_data = ['train'], test_data = ['test'], test_data_prop = 0.1)
```
The path should come from CLI args instead of hardcoding here

bhavul reviewed (commented):
Have checked caption_task, will have a look at vqa soon as well.
Need to try out code with debugger to make sure slicing is correct.

SasankYadati committed:
TextTask and ControlTask don't need explicit TaskType to be passed into each of them

SasankYadati committed:
evaluate text tasks

eihli committed:
Add newline and .egg-info to .gitignore

Reason for .egg-info: https://github.com/python-poetry/poetry/issues/2757

eihli committed:
Rename conda environment from gato-control to neko

Now that NEKO is doing more than just control tasks, a new name seems appropriate.

eihli committed:
Remove unused imports and fix type

eval.py passes a value for pretrained_lm that is possibly None, so
GatoPolicy needs to accept the value as optional.

eihli committed:
Refactor to not need TaskTypeEnum

Using isinstance gives us better type hinting.

nomitchell committed:
Changed saved model name from an ID to a timestamp

bhavul committed:
Merge pull request #65 from SasankYadati/sai/eval_text_task

Enable eval.py to work on text datasets

bhavul committed:
Merge pull request #55 from eihli/fix-version-dependency-issue

Fix version dependency issue #54

eihli committed:
Merge branch 'master' into stricter-typing

bhavul committed:
Merge pull request #35 from eihli/stricter-typing

Update TaskType references to use enum, not str

henryj18 committed:
Fix evaluate() function in caption_task.py and vqa_task.py to randomly pick examples

henryj18 committed:
Minor change of one line of code in predict_response, actually not really change. just reorder the variable to make it clearer for understanding,

henryj18 committed:
Change the predict_response function to be aligned with similar function of other tasks

eihli committed:
Merge pull request #70 from nomitchell/timestamp

Changed saved model name from an ID to a timestamp

henryj18 committed:
Fix some typo

henryj18 committed:
Merge branch 'master' into add_vqa

henryj18 commented:
Merge add_vqa to master
